### PR TITLE
Fix UB in atomic memory operations to memory locations with automatic storage

### PR DESCRIPTION
**Pull request by gonzalobg**
Before this PR, libcu++ introduces _undefined behavior_ into well-formed programs that use atomic memory operations on memory locations of automatic storage duration. This UB materializes in incorrect results being observable by [users, e.g., depending on the number of threads used (stackoverflow question)](https://stackoverflow.com/questions/75640847/what-exactly-does-cudaatomic-does/75645169). Since most of the examples in libcu++'s documentation use automatic variables, they all introduce this form of UB. 

This PR has two commits:
- [x] Adds a test for atomic automatic variables that fails when run with more than 1 thread per block: https://github.com/NVIDIA/libcudacxx/pull/427/commits/23f17f8d46d568cef197078ede1ad84740e5f49c
- [x] Fixes the codegen backend to test whether the state space of the memory location being operated on atomically is _local memory_, and if that's the case, fall back to a combination of weak memory operations (and others) to emulate local memory atomics, instead of like until now, generating PTX instructions for the _generic_ statespace which exhibit undefined behavior on local memory: https://github.com/NVIDIA/libcudacxx/pull/427/commits/2e4235e8cd65a5719b23883283f8351f0ade2f6d

Note: there is a third commit that disables the force include, since that whole mechanism seems to prevent anyone from writing tests with some "control" about threads involved, which seems to result in most of the tests for our atomic functionality to actually end up using only a single thread in practice.

### CODE REVIEW

**Review by miscco — CHANGES_REQUESTED** [review]
AFAIK the fake_main hack is used to run the tests both on device and on host. 

Disabling it will have unintended consequences of disabling device tests.

**Comment by gonzalobg** [comment]
Please advice on how to remove that commit. 

None of the atomic tests I found uses more than one thread, and here we need to use more than one thread to reproduce. 

I can remove the test and the workaround to support this test, and ship only the code change without tests.

**Comment by miscco** [comment]
> Please advice on how to remove that commit.
> 
> None of the atomic tests I found uses more than one thread, and here we need to use more than one thread to reproduce.
> 
> I can remove the test and the workaround to support this test, and ship only the code change without tests.

You need to have a look at the current tests for semaphore or the new PR for mutex. There are tests that involve multiple threads

File: include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_derived.h
Comment (jrhemstad) [inline include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_derived.h:16]: I assume this doesn't add overhead because the `__isLocal` check is optimized at compile time?

File: include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h
Comment (jrhemstad) [inline include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h:13]: This could/should use `__cuda_memcmp`. I'll leave it up to you to decide which implementation is better. 

https://github.com/NVIDIA/libcudacxx/blob/a13db9aeab643e01498e5fcc3f454bbb9ea1626c/include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda.h#L401-L419

File: include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_derived.h
Comment (gonzalobg) [reply include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_derived.h:16]: It will add overhead if it isn't optimized, but without it, you'll get wrong results faster.

There is a better patch we can do later were we recover this overhead, but that will require some more changes throughout the stack.

File: include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h
Comment (gonzalobg) [reply include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h:13]: They also suggested that they would prefer to expose memcmp through libcu++ instead of as an intrinsic, given that with that change it can be implemented reliably on the library side, so it might make sense to expose that through `cuda::std::memcmp` (if it isn't already).

File: .upstream-tests/test/cuda/atomics/atomic.local.pass.cpp
Comment (miscco) [inline .upstream-tests/test/cuda/atomics/atomic.local.pass.cpp:5]: This is missing our copyright

File: include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h
Comment (miscco) [inline include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_local.h:140]:
```suggestion
#endif // __LIBCUDACXX_ATOMIC_CUDA_LOCAL_H
```

**Review by miscco — APPROVED** [review]

**Review by wmaxey — CHANGES_REQUESTED** [review]
I'm blocking merge until we get the pipeline stuff solved. There's some weirdness happening that I haven't been able to solve. There's something very broken with the interleaved test in the producer consumer pipeline.

**Comment by gonzalobg** [comment]
> I'm blocking merge until we get the pipeline stuff solved. 

There is one test still failing with these changes: `cuda/pipeline_memcpy_async_producer_consumer.pass.cpp`; I can reproduce and will take a look.

**Comment by gonzalobg** [comment]
https://github.com/NVIDIA/cccl/pull/478

**Pull request closed without merging**
